kame/kame/shisad/cnd.8
.\"	$KAME: cnd.8,v 1.3 2005/05/28 22:49:43 keiichi Exp $
.\"
.\" Copyright (C) 2004 WIDE Project.
.\" All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the project nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd June 2, 2004
.Dt CND 8
.Os KAME
.Sh NAME
.Nm cnd
.Nd Mobile IPv6 correspondent node daemon
.\"
.Sh SYNOPSIS
.Nm
.Op Fl f
.Op Fl n
.Op Fl c Ar configfile
.\"
.Sh DESCRIPTION
.Nm
is a daemon to add a node Mobile IPv6 correspondent node capability, which maintains binding cache and handles MIP6 signaling packets.
Mobile IPv6 capable kernel is required to enable the function.
.Pp
The command line options are:
.Bl -tag -width indent
.\"
.It Fl f
Foreground mode.
.It Fl n
Never lookup hostname from a address on a telnet interface.
.It Fl c Ar configfile
Specify an alternative location for the configuration file. By default, 
.Pa /usr/local/v6/etc/cnd.conf
is used.
.El
.Pp
An intaractive telnet interface on port 7777 is provided to see it's binding cache.
The socket opens only for IPv6 local loopback. 
.Pp
The commands list will be displayed with '?' or 'help' on the prompt.
.\"
.Sh HISTORY
The
.Nm
command first appeared in Shisa Mobile IPv6 protocol stack.
.Sh BUGS
All of parts.


CHANGELOG
			CHANGELOG for KAME kit
$KAME: CHANGELOG,v 1.2769 2005/05/25 03:22:29 keiichi Exp $

<200505>
Wed May 25 12:20:25 JST 2005	keiichi@iijlab.net
	* kame/kame/shisa: now shisa daemons uses configuration files
	instead of command line options.  please check sample
	configuratoin files located in the shisa directory.

2005-05-19  SUZUKI, Shinsuke <suz@kame.net>
	* (freebsd5|netbsd)/sys/sys/sockio.h: assign the same ioctl number 
	to MSF-API-related ioctls among all *BSDs (to keep binary 
	compatibilities as much as possible)

2005-05-14  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/libinet6/getaddrinfo.c (matchlen): corrected a wrong
	cast from sockaddr_in to sockaddr_in6.

2005-05-13  SUZUKI, Shinsuke <suz@kame.net>
	* freebsd5: sync with 5.4-RELEASE (it includes a fix for
	the FreeBSD Security Advisory FreeBSD-SA-05:06-09)
	
2005-05-03  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/dhcp6c_ia.c (ia_timo): kept the latest server's
	DUID even in the REBIND state in case of sending a Release message
	in that state.

2005-05-03  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/cfparse.y: corrected "address parameters" so
	that we don't have to specify the meaningless (but mandatory)
	prefix length in the context of IA_NA.

2005-05-03  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/config.c (configure_commit): made sure that
	interface parameters are initialized with the default values even
	if the interface is not explicitly configured in the configuration
	file.  This particularly made sure that the preference option is
	not included unless explicitly specified.  A memory leakage for
	the client when specifying a script file was also fixed.

<200504>
2005-04-29  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/dhcp6c.c (client6_mainloop): corrected error
	handling for select(2) in order to avoid unexpected blocking when
	receiving a signal.

Thu Apr 28 16:12:38 JST 2005 sakane@tanu.org
	* freebsd[45]/usr.sbin/Makefile,netbsd/usr.sbin/Makefile:
	racoon was removed from the make list.

2005-04-28  SUZUKI, Shinsuke <suz@kame.net>
	* freebsd5/sbin/ip6fw: stop including ip6fw in SNAP, since FreeBSD's
	ip6fw is enough (KAME's ip6fw does not have any function different 
	from ip6fw)

2005-04-27  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/sys/netinet6/in6.c (in6_update_ifa): always updated
	ia6_updatetime regardless of whether the address is created or
	updated.  This is necessary so that the expiration times will be
	updated	correctly when the lifetiems of an existing address are
	modified by hand or by a process such as a DHCPv6 client.

2005-04-26  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/addrconf.c (na_ifaddrconf): corrected lifetime
	arguments to ifaddrconf().

2005-04-23  SUZUKI, Shinsuke <suz@kame.net>
	* kame/sys/netinet6/in6_rmx.c, in6_ifattach.c: fixed a kernel crash
	by a lock-order-reversal in freebsd5

2005-04-20  SUZUKI, Shinsuke <suz@kame.net>
	* freebsd5/sys/net/if.c: FreeBSD Security Advisory
	 FreeBSD-SA-05:04.ifconf

2005-04-19  SUZUKI, Shinsuke <suz@kame.net>
	* *bsd/sys/netinet/igmp.c, igmp_var.h, in.[ch], ip_output.c,
	  kame/sys/netient/in_msf.c: 
	  IGMPv3 treats 224.0.0.x(x!=1) as a normal IPv4 multicast address,
	  as specified in RFC3766.

	* openbsd/sys/netinet/in.c, in_var.h, ip_output.c,
	* netbsd/sys/netinet/igmp.c, igmp_var.h, in.c, in_pcb.c, in_var.h,
	  ip_output.c: reflect the following MLDv2 changes into the 
	  NetBSD/OpenBSD's IGMPv3 implementation.
	- fixs a kernel crash by an IGMPv3 Query (2005-01-01)
	- retransmits IGMPv3 report RobustnessVariable times (2004-12-31)
	- corrects an IGMP-query handling when MaxResponseCode=0 (2004-12-27)
	- introduces a sysctl option to force IGMP version (2004-07-09)
	- renames IGMPv3-ready in_multi{} manipulation functions (2004-02-05)

	* freebsd5/sys/netinet/igmp.c: fixed a kernel crash by an IGMP packet
	(a bug introduced in the change at 2005-04-14)

2005-04-18  SUZUKI, Shinsuke <suz@kame.net>
	* freebsd5/sys/netinet/igmp.c, in.c, ip_output.c,
	  kame/sys/netinet/in_msf.c: fixed the byte order of an IPv4 address 
	  for IN_XXX macros (freebsd-only)

2005-04-14  SUZUKI, Shinsuke <suz@kame.net>
	* kame/sys/netinet/in_msf.c freebsd5/contrib/pf/net/if_pfsync.c,
	  freebsd5/sys/netinet/igmp.c, igmp_var.h, in.c, in_pcb.c, in_var.h,
	  ip_output.c: reflect the following MLDv2 changes into the FreeBSD5's
	  IGMPv3 implementation.
	- fixs a kernel crash by an IGMPv3 Query (2005-01-01)
	- retransmits IGMPv3 report RobustnessVariable times (2004-12-31)
	- corrects an IGMP-query handling when MaxResponseCode=0 (2004-12-27)
	- introduces a sysctl option to force IGMP version (2004-07-09)
	- renames IGMPv3-ready in_multi{} manipulation functions (2004-02-05)

	* freebsd5/sys/netinet/ip_output.c: specifies spl()s as done in 
	  original FreeBSD5

	* kame/*: remove freebsd[234]-specific code, since they are no longer
	  supported in KAME-SNAP

2005-04-05  SUZUKI, Shinsuke <suz@kame.net>
	* freebsd5/sys/ufs/ffs/ffs_inode.c: FreeBSD Security Advisory
	 FreeBSD-SA-05:02.sendfile

2005-04-04  SUZUKI, Shinsuke <suz@kame.net>
	* kame/sys/netinet6/nd6.c: when the number of the packets queued in
	  a NDP cache exceed the limit, discard the oldest packet, 
	  as specified in RFC2461 7.2.2.

2005-04-01  SUZUKI, Shinsuke <suz@kame.net>
	* kame/sys/netinet/icmp6.h, kame/sys/netinet6/icmp6.c, nd6.c: limits 
	the number of unresolved packets stored in a NDP cache.
	(To prevent a DoS attack using 2005-01-21's change)

2005-04-01  SUZUKI, Shinsuke <suz@kame.net>
	* kame/kame/rtadvd/configc, rtadvd.[ch]: reflect router-variables
	in RA configuration into host-variables in kernel.

<200503>
2005-03-30  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/addrconf.c (update_address): make sure that
	the lifetimes of addresses are updated when the client receives a
	Reply in response to Renew or Rebind.


kame/kame/shisad/cnd.c
/*	$KAME: cnd.c,v 1.11 2005/05/24 10:16:19 keiichi Exp $	*/

/*
 * Copyright (C) 2004 WIDE Project.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the project nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include <syslog.h>
#include <poll.h>

#include <sys/param.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/sockio.h>

#include <net/if.h>
#ifdef __FreeBSD__
#include <net/if_var.h>
#endif
#include <netinet/in.h>
#include <netinet/icmp6.h>
#include <netinet/ip6mh.h>
#include <net/mipsock.h>
#include <netinet6/mip6.h>

#include "callout.h"
#include "stat.h"
#include "shisad.h"
#include "fdlist.h"
#include "command.h"
#include "config.h"

/*static void command_show_status(int, char *);*/
static void command_flush(int, char *);
static void terminate(int);

struct command_table show_command_table[] = {
	{"bc", command_show_bc, "binding chache"},
	{"kbc", command_show_kbc, "binding chache in kernel"},
	{"stat", command_show_stat, "statisticts"},
	{"callout", show_callout_table, "show callout table "},
	{NULL}
};

struct command_table command_table[] = {
	{"show", NULL, "Show status", show_command_table},
	{"flush", command_flush, "Flush binding caches"},
};

/* Global Variables */
int mhsock, mipsock, icmp6sock;
struct mip6stat mip6stat;
int homeagent_mode = 0;

/* configuration parameters */
int debug = 0;
int foreground = 0;
int namelookup = 1;
int command_port = CND_COMMAND_PORT;
char *conffile = CND_CONFFILE;

static void cn_lists_init(void);
void flush_bc(void);

void
cn_usage(path)
	char *path;
{
	char *cmd;

	cmd = strrchr(path, '/');
	if (!cmd)
		cmd = path;
	else
		cmd++;
	fprintf(stderr, "%s [-fn] [-c configfile]\n", cmd);
	exit(0);
} 


int
main(argc, argv)
	int argc;
	char **argv;
{
	int pfds;
	int ch = 0;
	FILE *pidfp;

	/* get options */
	while ((ch = getopt(argc, argv, "fnc:")) != -1) {
		switch (ch) {
		case 'f':
			foreground = 1;
			break;
		case 'n':
			namelookup = 0;
			break;
		case 'c':
			conffile = optarg;
			break;
		default:
			fprintf(stderr, "unknown option\n");
			cn_usage();
			break;
		}
	}

	/* parse configuration file and set default values. */
	if (parse_config(CFM_CND, conffile) == 0) {
		config_get_number(CFT_DEBUG, &debug, config_params);
		config_get_number(CFT_COMMANDPORT, &command_port,
		    config_params);
	}

	/* open syslog infomation. */
	openlog("shisad(cnd)", 0, LOG_DAEMON);
	syslog(LOG_INFO, "-- Start CN daemon at -- \n");

	/* open sockets */
	mhsock_open();
	mipsock_open();
	icmp6sock_open();

	/* start timer */
	callout_init();

	/* initialization */
	fdlist_init();
	if (command_init("cn> ", command_table, sizeof(command_table) / sizeof(struct command_table), command_port) < 0) {
		fprintf(stderr, "Unable to open user interface\n");
	}
	cn_lists_init();
	init_nonces();

	/* register signal handlers. */
	signal(SIGTERM, terminate);
	signal(SIGINT, terminate);

	if (foreground == 0)
		daemon(0, 0);

	/* dump current PID */
	if ((pidfp = fopen(CND_PIDFILE, "w")) != NULL) {
		fprintf(pidfp, "%d\n", getpid());
		fclose(pidfp);
	}

	new_fd_list(mhsock, POLLIN, mh_input_common);
        new_fd_list(mipsock, POLLIN, mipsock_input_common);
        new_fd_list(icmp6sock, POLLIN, icmp6_input_common);

	/* notify a kernel to behave as a correspondent node. */
	mipsock_nodetype_request(MIP6_NODETYPE_CORRESPONDENT_NODE, 1);

	while (1) {
		clear_revents();
	    
		if ((pfds = poll(fdl_fds, fdl_nfds, get_next_timeout())) < 0) {
			perror("poll");
			continue;
		}

		if (pfds != 0) {
			dispatch_fdfunctions(fdl_fds, fdl_nfds);
		}
		/* Timeout */
		callout_expire_check();
	}

	return (0);
}

static void
cn_lists_init(void)
{
	mip6_bc_init();

	return;
}


int
mipsock_input(miphdr)
	struct mip_msghdr *miphdr;
{
	int err = 0;
	struct mipm_nodetype_info *nodeinfo;

	switch (miphdr->miph_type) {
	case MIPM_NODETYPE_INFO:
		nodeinfo = (struct mipm_nodetype_info *)miphdr;
		homeagent_mode = nodeinfo->mipmni_enable & MIP6_NODETYPE_HOME_AGENT;
	case MIPM_BE_HINT:
		mipsock_behint_input(miphdr);
		break;
	default:
		break;
	}

	return (err);
}

static void
command_flush(s, arg)
	int s;
	char *arg;
{
	flush_bc();
}

static void
terminate(dummy)
	int dummy;
{
	mip6_flush_kernel_bc();
	mipsock_nodetype_request(MIP6_NODETYPE_CORRESPONDENT_NODE, 0);
	unlink(CND_PIDFILE);
	exit(1);
}
